### ISSUE TITLE
Container falling over: failed to find required state file: /data/25232264-75d098483a.lstate

### ISSUE THREAD

**Issue by kss-espeo** [input-output-hk/cardano-db-sync#563]
System spec:
```
# uname -a
Linux cardano-node-01 5.4.0-47-generic #51~18.04.1-Ubuntu SMP Sat Sep 5 14:35:50 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.4 LTS
Release:	18.04
Codename:	bionic
docker-compose/bionic,now 1.17.1-2 all [installed]
docker.io/bionic-updates,now 19.03.6-0ubuntu1~18.04.3 amd64 [installed]
inputoutput/cardano-explorer-api:3.1.2
inputoutput/cardano-submit-api:3.1.2
inputoutput/cardano-db-sync:8.0.0
inputoutput/cardano-graphql-hasura:3.2.0
inputoutput/cardano-graphql:3.2.0
postgres:11.5-alpine
inputoutput/cardano-node:1.25.1

Cardano stack installed following:
https://github.com/input-output-hk/cardano-graphql
```

Hi. I am running a Cardano Node along with a GraphQL interface like this:

```
krzysztof-spisak-spisacki@cardano-node-01:~$ sudo docker ps -a
CONTAINER ID        IMAGE                                      COMMAND                  CREATED             STATUS                          PORTS                              NAMES
a438a9524e0a        inputoutput/cardano-explorer-api:3.1.2     "/nix/store/yvgr5ify…"   3 weeks ago         Up 3 hours                      0.0.0.0:8100->8100/tcp             cardanorest_cardano-explorer-api_1
084b7d0a08d8        inputoutput/cardano-submit-api:3.1.2       "/nix/store/9fx3jhv2…"   3 weeks ago         Up 3 hours                      8090/tcp, 0.0.0.0:8091->8091/tcp   cardanorest_cardano-submit-api_1
8a7d7bbf23fd        inputoutput/cardano-db-sync:8.0.0          "/nix/store/3c10spnd…"   3 weeks ago         Restarting (1) 35 seconds ago                                      cardanographql_cardano-db-sync-extended_1
f44e93fdb138        inputoutput/cardano-graphql-hasura:3.2.0   "docker-entrypoint.s…"   3 weeks ago         Up 3 hours                      0.0.0.0:8090->8080/tcp             cardanographql_hasura_1
d105da663bbf        inputoutput/cardano-graphql:3.2.0          "node index.js"          3 weeks ago         Up 3 hours                      0.0.0.0:3100->3100/tcp             cardanographql_cardano-graphql_1
1aab79eec668        postgres:11.5-alpine                       "docker-entrypoint.s…"   3 weeks ago         Up 3 hours                      0.0.0.0:5432->5432/tcp             cardanographql_postgres_1
8cbbd680eb78        inputoutput/cardano-node:1.25.1            "/nix/store/jmscm2vc…"   3 weeks ago         Up 3 hours                                                         cardanographql_cardano-node_1
```

I have been running it for 2 weeks with no issues. However since recently, the `cardano-db-sync` component tends to fall over before bringing itself back to life. It then might instantly fall over again, operate for an extended period of time before inevitably falling over again. Find a tail of log produced before container exits below:

```
Connecting to network specified in configuration.yaml
[db-sync-node:Info:4] [2021-03-30 12:37:37.42 UTC] NetworkMagic: 764824073
[db-sync-node:Info:4] [2021-03-30 12:37:37.46 UTC] Initial genesis distribution present and correct
[db-sync-node:Info:4] [2021-03-30 12:37:37.46 UTC] Total genesis supply of Ada: 31112484745.000000
[db-sync-node:Info:4] [2021-03-30 12:37:37.47 UTC] Inserting Shelley Genesis distribution
[db-sync-node:Info:4] [2021-03-30 12:37:37.47 UTC] epochPluginOnStartup: Checking
[db-sync-node:Info:4] [2021-03-30 12:37:42.97 UTC] localInitiatorNetworkApplication: connecting to node via "/node-ipc/node.socket"
[db-sync-node.Subscription:Notice:42] [2021-03-30 12:37:42.97 UTC] Identity Starting Subscription Worker, valency 1
[db-sync-node.Subscription:Notice:43] [2021-03-30 12:37:42.97 UTC] Identity Connection Attempt Start, destination LocalAddress "/node-ipc/node.socket"
[db-sync-node.Subscription:Notice:43] [2021-03-30 12:37:42.97 UTC] Identity Connection Attempt End, destination LocalAddress "/node-ipc/node.socket" outcome: ConnectSuccessLast
[db-sync-node.Handshake:Info:43] [2021-03-30 12:37:42.97 UTC] WithMuxBearer (ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"}) Send (ClientAgency TokPropose,MsgProposeVersions (fromList [(NodeToClientV_1,TInt 764824073),(NodeToClientV_2,TInt 764824073),(NodeToClientV_3,TInt 764824073),(NodeToClientV_4,TInt 764824073),(NodeToClientV_5,TInt 764824073),(NodeToClientV_6,TInt 764824073),(NodeToClientV_7,TInt 764824073),(NodeToClientV_8,TInt 764824073)]))
[db-sync-node.Handshake:Info:43] [2021-03-30 12:37:42.97 UTC] WithMuxBearer (ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"}) Recv (ServerAgency TokConfirm,MsgAcceptVersion NodeToClientV_8 (TInt 764824073))
[db-sync-node.Mux:Info:43] [2021-03-30 12:37:42.97 UTC] Bearer on ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"} event: Handshake Client end, duration 0.001405729s
[db-sync-node.Mux:Info:45] [2021-03-30 12:37:42.97 UTC] Bearer on ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"} event: State: Mature
[db-sync-node:Info:48] [2021-03-30 12:37:42.97 UTC] Starting chainSyncClient
[db-sync-node:Info:48] [2021-03-30 12:37:42.99 UTC] Database tip is at slot 25232264, block 5512207
[db-sync-node:Info:52] [2021-03-30 12:37:42.99 UTC] Running DB thread
[db-sync-node.Subscription:Notice:42] [2021-03-30 12:37:43.02 UTC] Identity Required subscriptions started
[db-sync-node:Info:52] [2021-03-30 12:37:44.61 UTC] Rolling back to slot 25232264, hash 75d098483a5c237c470dfd993cdd10096fc55147b5f910476c5881edcf4569ec
[db-sync-node:Info:52] [2021-03-30 12:37:44.61 UTC] Deleting slots numbered: []
[db-sync-node:Info:52] [2021-03-30 12:37:44.61 UTC] Slots deleted
[db-sync-node:Error:52] [2021-03-30 12:37:44.67 UTC] runDBThread: FatalError {fatalErrorMessage = "loadLedgerStateAtPoint failed to find required state file: /data/25232264-75d098483a.lstate"}
[db-sync-node:Error:55] [2021-03-30 12:37:44.69 UTC] recvMsgRollForward: AsyncCancelled
[db-sync-node:Error:48] [2021-03-30 12:37:44.69 UTC] ChainSyncWithBlocksPtcl: FatalError {fatalErrorMessage = "loadLedgerStateAtPoint failed to find required state file: /data/25232264-75d098483a.lstate"}
[db-sync-node.Mux:Info:45] [2021-03-30 12:37:44.69 UTC] Bearer on ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"} event: State: Dead
[db-sync-node.Mux:Notice:45] [2021-03-30 12:37:44.69 UTC] Bearer on ConnectionId {localAddress = LocalAddress "", remoteAddress = LocalAddress "/ipc/node.socket"} event: Miniprotocol MiniProtocolNum 5 InitiatorDir terminated with exception FatalError {fatalErrorMessage = "loadLedgerStateAtPoint failed to find required state file: /data/25232264-75d098483a.lstate"}
[db-sync-node.Subscription:Error:43] [2021-03-30 12:37:44.70 UTC] Identity Application Exception: LocalAddress "/node-ipc/node.socket" FatalError {fatalErrorMessage = "loadLedgerStateAtPoint failed to find required state file: /data/25232264-75d098483a.lstate"}
[db-sync-node.ErrorPolicy:Error:4] [2021-03-30 12:37:44.70 UTC] IP LocalAddress "/node-ipc/node.socket" ErrorPolicyLocalNodeError (ApplicationExceptionTrace (FatalError {fatalErrorMessage = "loadLedgerStateAtPoint failed to find required state file: /data/25232264-75d098483a.lstate"}))
cardano-db-sync-extended: FatalError {fatalErrorMessage = "loadLedgerStateAtPoint failed to find required state file: /data/25232264-75d098483a.lstate"}

```

Any idea what's happening? Has the state of db somehow become corrupted? What is the course of action to take in order to recover entire stack, so that GraphQL interface is usable again? Also - how did it end up in such state and how to avoid this happening in the future?

Thanks,
KSS

**Comment by kss-espeo** [comment]
> `cardano-db-sync` has minimum [system requirements](https://github.com/input-output-hk/cardano-db-sync#system-requirements) . Are you meeting them?
Yes, I am meeting all the requiremenets listed under this link.
> Also that `/data/` directory needs to be persistent across reboots.
I assume you mean machine reboot. Yes, `/data` is being persisted across reboots.

**Comment by erikd** [comment]
I run `db-sync` all the time and I have not seen this issue since I first implemented and debugged this feature.

What is actually in the `/data`  directory? My suspicion is that the `--state-dir` parameter you passed last time you ran it is different from the one you passed this time and that is why it failed to find any ledger state tiles.

Maybe @rhyslbw can help with this one.

**Comment by kss-espeo** [comment]
Contents of `/data` on `db-sync` container:

```
bash-4.4# ls /data
0-67656e6573.lstate	    25232111-af4eb78ad8.lstate	25232153-c3e74d14d1.lstate  25232189-6b4d3f9e0a.lstate	25232264-d43ececf4a.lstate
25232110-3fd978f75e.lstate  25232140-c2624c1a8c.lstate	25232180-aeb0382ab9.lstate  25232197-5da616738e.lstate
```

**Comment by rhyslbw** [comment]
@erikd It's looking for `25232264-75d098483a.lstate`, which I can see is the accepted block. Stored on disk here is `25232264-d43ececf4a.lstate`, which is presumably one that was rolled back.

**Comment by erikd** [comment]
Does it work, it the latest ledger state is deleted? The latest one is the one with the highest numbered `X` in `X-Y.lstate`

**Comment by kss-espeo** [comment]
It seems it has recovered after some time. Does it mean my node somehow did not process a rollback? Any idea how can I prevent this from happening again?

**Comment by kderme** [comment]
Thanks for reporting this. It seems there is an issue with how db-sync handles these files in case of rollbacks. We plan to improve this and allow db-sync to better self-heal in cases like this.

In the meantime a workaround as @erikd said is to manually delete the newest ledger state file and restart db-sync.

**Comment by erikd** [comment]
I have now hit this three times in the last 12 hours on my local instance.

**Comment by kss-espeo** [comment]
I am sorry for cursing the network.

**Comment by kss-espeo** [comment]
After some time same issues happened again. It complains about lack of another file:
```
cardano-db-sync-extended: FatalError {fatalErrorMessage = "loadLedgerStateAtPoint failed to find required state file: /data/25837734-489201a10c.lstate"}
```

Current state of `/data`:
```
0-67656e6573.lstate             25837683-e7cb5b6ae8.lstate  25837705-f29d0395cc.lstate  25837725-cb2a090f0f.lstate  25837730-89e53f5d80.lstate
25232264-d43ececf4a.lstate.old  25837686-658e5b4522.lstate  25837706-26d7dcb873.lstate  25837726-78a7328303.lstate  25837734-c6fe56eec9.lstate
```

**Comment by kss-espeo** [comment]
After removing most recent file again, I left the node in peace and it seemed to be fine for few days.
Then I started my app which resulted in multiple calls to node's GraphQL API (approx 20-30 times per second) . Shortly after that, an issue describerd above reappeared.

**Comment by erikd** [comment]
This was probably fixed in #584 on `master`. There is likely to be a new release (which will include this fix) either late this week or early next week.

**Comment by erikd** [comment]
The release with the fix for #585 should land this week.

**Comment by erikd** [comment]
Release `10.0.0` has been tagged and the official release will occur after some more QA.
